### PR TITLE
SparseState spin2

### PR DESCRIPTION
**Pull request by brianz98**
## Description
Add ability to calculate <Psi0|S^2|Psi1> where Psi0 and Psi1 are `SparseState` objects. Also normalize (global function) function for `SparseState`.

## Checklist
- [x] Added/updated tests of new features and included a reference `output.ref` file
- [x] Ready to go!

### CODE REVIEW

File: forte/api/sparse_state_api.cc [hunk 1]
```diff
@@ -73,6 +73,8 @@ void export_SparseState(py::module& m) {
 
     m.def("apply_number_projector", &apply_number_projector);
     m.def("get_projection", &get_projection);
+    m.def("spin2_sparse", &spin2_sparse);
```
Comment (EmilyHuang2005) [inline forte/api/sparse_state_api.cc:76]: Please add a comment.

File: forte/api/sparse_state_api.cc [hunk 2]
```diff
@@ -73,6 +73,8 @@ void export_SparseState(py::module& m) {
...
     m.def("overlap", &overlap);
+    m.def("normalize", &normalize);
```
Comment (wateriousmelon2333) [inline forte/api/sparse_state_api.cc:78]:
```suggestion
    m.def("normalize", &normalize); 
    returns a normalized version of the import sparse state 
```

File: forte/sparse_ci/sparse_state.cc [hunk 3]
```diff
@@ -185,6 +185,26 @@ sparse_scalar_t overlap(const SparseState& left_state, const SparseState& right_
     return left_state.dot(right_state);
 }
 
+sparse_scalar_t spin2_sparse(const SparseState& left_state, const SparseState& right_state) {
+    sparse_scalar_t s2 = 0.0;
+    for (const auto& [deti, ci] : left_state) {
+        for (const auto& [detj, cj] : right_state) {
+            s2 += ci * cj * spin2(deti, detj);
```
Comment (fevangelista) [inline forte/sparse_ci/sparse_state.cc:192]:
```suggestion
            s2 += conjugate(ci) * cj * spin2(deti, detj);
```

File: forte/sparse_ci/sparse_state.cc [hunk 4]
```diff
@@ -185,6 +185,26 @@ sparse_scalar_t overlap(const SparseState& left_state, const SparseState& right_
...
+        }
+    }
+    return s2;
+}
+
+SparseState normalize(const SparseState& state) {
+    SparseState new_state;
+    const auto n = state.norm();
+    for (const auto& [det, c] : state) {
```
Comment (fevangelista) [inline forte/sparse_ci/sparse_state.cc:201]:
```suggestion
     SparseState new_state(state);
     new_state /= n;
```

File: forte/sparse_ci/sparse_state.h [hunk 5]
```diff
@@ -73,4 +73,9 @@ SparseState apply_number_projector(int na, int nb, const SparseState& state);
 /// compute the overlap value <left_state|right_state>
 sparse_scalar_t overlap(const SparseState& left_state, const SparseState& right_state);
 
+/// compute the S^2 expectation value
+sparse_scalar_t spin2_sparse(const SparseState& left_state, const SparseState& right_state);
+
+SparseState normalize(const SparseState& state);
```
Comment (fevangelista) [inline forte/sparse_ci/sparse_state.h:79]: Add comment above

**Review by fevangelista — APPROVED** [review]
Thank you for making this PR!

**Pull request merged**